### DISCUSSION TITLE
Headless Screenshot with for-loop [assistance/help]

### DISCUSSION THREAD

**Discussion by engineertree5** [highfestiva/finplot#77] [Q&A] [answered]
Hey Highfestiva,

I originally asked my question under an issue and realized that was the wrong place to ask a question. I think the discussion board is the best place to do this. hoping my novice question will help others and myself better understand python and your finplot app. 

Anyways, I'm still trying to figure out how your screenshot() function works. Let me explain what I understand is happening and maybe you or someone can correct me where I'm wrong.

> GOAL: Loop through watchlist and save charts in headless mode. Testing on local laptop to ensure everything is working as expcted. 

working in headless mode was shared [here](https://github.com/highfestiva/finplot/issues/28). Which I copied this example but failed to execute. I was still having to close out of my chart to proceed through the for-loop. also passing variables into `fplt.timer_callback(save(variable), 0.5, single_shot=True)` was throwing errors. That is not point, least I dont think it was the root of my probmels... carrying on. 

# Breaking down screenshot():

starting at line [1598](https://github.com/highfestiva/finplot/blob/448bd43c6c2c7828d8b5a1a8cddf223c5984eb68/finplot/__init__.py#L1598-L1609) You must use the fplt.timer_callback() function before you can use the screenshot function. IF the call back function is used, move forward (go into `try:` statement) and grab a screenshot of the primary window. 

```python
def screenshot(file, fmt='png'):
    if _internal_windows_only() and not app:
        print('ERROR: save_screenshot must be callbacked from e.g. timer_callback()')
        return False
    try:
        buffer = QtCore.QBuffer()
        app.primaryScreen().grabWindow(windows[0].winId()).save(buffer, fmt)
        file.write(buffer.data())
        return True
    except Exception as e:
        print(type(e), e)
    return False
```

This all makes sense to me and this feature is working. However, when i'm testing this function inside a for-loop I have to exit the window each time. is it possible to save a chart without using the `fplt.show()` function?
 
lastly, I'm familiar with taking a screenshot in headless mode. you shared how to do so [here](https://github.com/highfestiva/finplot/issues/28). when using `io.BytesIO` you are saveing the screenshot to memory. then letting the user do "something" with it, assuming save it. but when testing I'm still having to close out the window from the `fplt.show()` function. 

is it my misunderstanding of how to save an image?  matplotlib has an option called [`savefig`](https://matplotlib.org/3.3.3/api/_as_gen/matplotlib.pyplot.savefig.html) which lets you save the figure without the need to show it. my forloop works with this but your finplot graphs/charts as so CLEAN I want to use finplot over matplot. 

Below is my code which WORKS but I have to exit each time to let the for-loop continue. If possible I'd like to not interact with the computer during proccessing stocks. 

```python

#!/usr/local/bin/python3

import finplot as fplt
import numpy as np
import pandas as pd
import datetime as dt
import pandas_datareader.data as web
import io

company_symbol = ['SHOP', 'NET']

today = dt.date.today()
end = dt.datetime(today.year, today.month,today.day)
start = dt.datetime(today.year -1, today.month,today.day)

def save():
    # f = io.BytesIO()
    # fplt.screenshot(f)
    # do_something(f)
    fplt.screenshot(open('screenshot.png', 'wb'))

for symbol in company_symbol:

    # df = web.DataReader(f'{symbol}', 'yahoo', start=start, end=end)
    # df.to_csv(f'{symbol}.csv')

    df = pd.read_csv(f'{symbol}.csv', parse_dates=True, index_col=0)
    fplt.candlestick_ochl(df[['Open','Close','High','Low']])
    # ax = fplt.create_plot(f'{symbol}', rows=1)

    fplt.timer_callback(save, 0.5, single_shot=True) # wait some until we're rendered
    fplt.show()
```

**Comment by highfestiva** [comment, marked as answer]
You only have to `show()` once, then use the (timer) callbacks to update and take screenshots. Update is done through `.update_data()`, see examples.

**Reply by wgong** [reply]
Hi, The candlestick charts created by your finplot are better looking than all the others - https://coderzcolumn.com/tutorials/data-science/candlestick-chart-in-python-mplfinance-plotly-bokeh
However, I understand finplot is originally designed for interactive app, In order to create chart png file, the current way of doing screenshot and callback seems to be a hack to me. Can you share your design suggestion for a completely headless feature (similar to matplotlib/savefig) ? Thank you.

**Reply by highfestiva** [reply]
finplot uses Qt5 under the hood. I do not believe there is an easy way to go "completely headless," and you're right that it's more of a hack. This feature request is on par with web rendering, and is out of scope for finplot. The overarching principle for finplot is to simplify backtesting and TA in python. (Anything else is a bonus.)

**Reply by wgong** [reply]
Appreciate your reply. Going to use finplot for interactive charting, and use matplotlib to generate chart image

**Reply by highfestiva** [reply]
Quick heads-up: in the most recent version, finplot can reset an axis and allow you to draw something new on it. See for instance `examples/complicated.py` where I reset the axes when changing asset, time span, etc.
